theano/sandbox/gpuarray/corr_gemm.cl

```diff
@@ -0,0 +1,132 @@
+/*
+Copyright (c) 2014, The Regents of the University of California (Regents)
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions are met: 
+
+1. Redistributions of source code must retain the above copyright notice, this
+   list of conditions and the following disclaimer. 
+2. Redistributions in binary form must reproduce the above copyright notice,
+   this list of conditions and the following disclaimer in the documentation
+   and/or other materials provided with the distribution. 
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+#undef _GLIBCXX_ATOMIC_BUILTINS
+
+__kernel void im2col_kernel(const int n, const __global float4 * data_im,
+    const int height, const int width, const int kernel_h, const int kernel_w,
+    const int pad_h, const int pad_w,
+    const int stride_h, const int stride_w,
+    const int height_col, const int width_col,
+    __global float4 * data_col) 
+{
+    for (int i = get_global_id(0); i < (n); i += get_global_size(0)) 
+    {
+    	int w_out = i % width_col;
+    	int h_index = i / width_col;
+    	int h_out = h_index % height_col;
+    	int channel_in = h_index / height_col;
+    	int channel_out = channel_in * kernel_h * kernel_w;
+    	int h_in = h_out * stride_h - pad_h;
+    	int w_in = w_out * stride_w - pad_w;
+    	float4 * data_col_ptr = data_col;
+    	data_col_ptr += (channel_out * height_col + h_out) * width_col + w_out;
+    	const float4 * data_im_ptr = data_im;
+    	data_im_ptr += (channel_in * height + h_in) * width + w_in;
+    	for (int i = 0; i < kernel_h; ++i) 
+	{
+      	    for (int j = 0; j < kernel_w; ++j) 
+	    {
+        	int h = h_in + i;
+        	int w = w_in + j;
+        	*data_col_ptr = (h >= 0 && w >= 0 && h < height && w < width) ?
+            	data_im_ptr[i * width + j] : 0;
+        	data_col_ptr += height_col * width_col;
+      	    }
+    	}
+    	barrier(CLK_LOCAL_MEM_FENCE | CLK_GLOBAL_MEM_FENCE);
+    }
+}
+
+void im2col(const __global float4 * data_im, const int channels,
```

Comment:
Or rather, the porting will have to do something different since it appears you tried to use cuda-style inline launches which is not supported by OpenCL.  The code that calls the kernel has to be separate from the code that is the kernel.